src/flows/context/results.tsx

```diff
@@ -32,41 +33,85 @@ export const ResultsProvider: FC = ({children}) => {
   const [statuses, setStatuses] = useState<Hash<RemoteDataState>>({})
   const [results, setResults] = useState({...EMPTY_STATE})
 
+  const updates = useRef([])
+  const [inc, setInc] = useState(0)
+
   useEffect(() => {
     setResults({...EMPTY_STATE})
   }, [flow?.id])
 
+  useEffect(() => {
+    if (!updates.current.length) {
+      return
+    }
+
+    let sChange = false
+    let rChange = false
+
+    while (updates.current.length) {
+      const [t, id, val] = updates.current.pop()
+      if (t === 'status') {
+        statuses[id] = val
+        sChange = true
+      }
+
+      if (t === 'result') {
+        results[id] = val
+        rChange = true
+      }
+    }
+
+    if (rChange) {
+      setResults({...results})
+    }
+
+    if (sChange) {
+      setStatuses({...statuses})
+    }
+
+    setInc(inc + 1)
+  }, [inc])
```

Comment:
this feels super wrong. I get why you're doing it, but it's inherently embedding another layer of complexity where we're now using state to manage state setting? That's at least what it feels like. I know why you're doing it, I just don't like it.